File: bonsai_adapter_ros_clf/src/main/java/de/unibi/citec/clf/bonsai/ros/actuators/RosTrackingActuator.java

package de.unibi.citec.clf.bonsai.ros.actuators;

import de.unibi.citec.clf.bonsai.actuators.TrackingActuator;
import de.unibi.citec.clf.bonsai.core.configuration.IObjectConfigurator;
import de.unibi.citec.clf.bonsai.core.exception.ConfigurationException;
import de.unibi.citec.clf.bonsai.ros.RosNode;
import de.unibi.citec.clf.bonsai.ros.helper.ResponseFuture;
import clf_perception_vision_msgs.ToggleCFtldTrackingWithBB;
import clf_perception_vision_msgs.ToggleCFtldTrackingWithBBRequest;
import clf_perception_vision_msgs.ToggleCFtldTrackingWithBBResponse;
import org.ros.exception.RosRuntimeException;
import org.ros.exception.ServiceNotFoundException;
import org.ros.namespace.GraphName;
import org.ros.node.ConnectedNode;
import org.ros.node.service.ServiceClient;
import java.util.List;
import java.util.concurrent.Future;

/**
 *
 * @author jkummert
 */
public class RosTrackingActuator extends RosNode implements TrackingActuator {

    String topic;

    private GraphName nodeName;
    private org.apache.log4j.Logger logger = org.apache.log4j.Logger.getLogger(getClass());

    private ServiceClient<ToggleCFtldTrackingWithBBRequest, ToggleCFtldTrackingWithBBResponse> clientTrigger;

    public RosTrackingActuator(GraphName gn) {
        initialized = false;
        this.nodeName = gn;
    }

    @Override
    public void configure(IObjectConfigurator conf) throws ConfigurationException {
        this.topic = conf.requestValue("topic");
    }

    @Override
    public GraphName getDefaultNodeName() {
        return nodeName;
    }

    @Override
    public void onStart(final ConnectedNode connectedNode) {
        try {
            clientTrigger           = connectedNode.newServiceClient(topic, ToggleCFtldTrackingWithBB._TYPE);
        } catch (ServiceNotFoundException e) {
            throw new RosRuntimeException(e);
        }
        initialized = true;
    }

    @Override
    public void destroyNode() {
        if(clientTrigger!=null) clientTrigger.shutdown();
    }

    @Override
    public Future<Boolean> startTracking(List<Integer> boundingbox){
        final ToggleCFtldTrackingWithBBRequest req = clientTrigger.newMessage();
        //set data
        req.getRoi().setXOffset(boundingbox.get(0));
        req.getRoi().setYOffset(boundingbox.get(1));
        req.getRoi().setHeight(boundingbox.get(2));
        req.getRoi().setWidth(boundingbox.get(3));
        final ResponseFuture<ToggleCFtldTrackingWithBBResponse> res = new ResponseFuture<>();
        clientTrigger.call(req, res);
        return res.toBooleanFuture();
    }

    @Override
    public void stopTracking(){
        final ToggleCFtldTrackingWithBBRequest req = clientTrigger.newMessage();
        //dont set data
        final ResponseFuture<ToggleCFtldTrackingWithBBResponse> res = new ResponseFuture<>();
        clientTrigger.call(req, res);
    }
}

File: bonsai_adapter_ros_clf/src/main/java/de/unibi/citec/clf/bonsai/ros/actuators/RosTrackingActuator.kt
package de.unibi.citec.clf.bonsai.ros.actuators

import clf_perception_vision_msgs.ToggleCFtldTrackingWithBB
import clf_perception_vision_msgs.ToggleCFtldTrackingWithBBRequest
import clf_perception_vision_msgs.ToggleCFtldTrackingWithBBResponse
import de.unibi.citec.clf.bonsai.actuators.TrackingActuator
import de.unibi.citec.clf.bonsai.core.configuration.IObjectConfigurator
import de.unibi.citec.clf.bonsai.core.exception.ConfigurationException
import de.unibi.citec.clf.bonsai.ros.RosNode
import de.unibi.citec.clf.bonsai.ros.helper.ResponseFuture
import de.unibi.citec.clf.btl.data.geometry.Point3D
import de.unibi.citec.clf.btl.data.geometry.Pose3D
import org.apache.log4j.Logger
import org.ros.exception.RosRuntimeException
import org.ros.exception.ServiceNotFoundException
import org.ros.namespace.GraphName
import org.ros.node.ConnectedNode
import org.ros.node.service.ServiceClient
import java.util.concurrent.Future

/**
 *
 * @author jkummert
 */
class RosTrackingActuator(gn: GraphName) : RosNode(), TrackingActuator {
    private lateinit var topic: String
    private val nodeName: GraphName
    private val logger = Logger.getLogger(javaClass)
    private var clientTrigger: ServiceClient<ToggleCFtldTrackingWithBBRequest, ToggleCFtldTrackingWithBBResponse>? =
        null

    init {
        initialized = false
        nodeName = gn
    }

    override fun configure(conf: IObjectConfigurator) {
        topic = conf.requestValue("topic")
    }

    override fun getDefaultNodeName(): GraphName {
        return nodeName
    }


    override fun connectionsAlive(): Boolean {
        return clientTrigger?.isConnected ?: false;
    }

    override fun onStart(connectedNode: ConnectedNode) {
        clientTrigger = try {
            connectedNode.newServiceClient(topic, ToggleCFtldTrackingWithBB._TYPE)
        } catch (e: ServiceNotFoundException) {
            throw RosRuntimeException(e)
        }
        initialized = true
    }

    override fun destroyNode() {
        if (clientTrigger != null) clientTrigger!!.shutdown()
    }

    override fun startTracking(boundingbox: List<Int>): Future<Boolean> {
        clientTrigger?.let { client ->
            val req = client.newMessage()
            //set data
            req.roi.xOffset = boundingbox[0]
            req.roi.yOffset = boundingbox[1]
            req.roi.height = boundingbox[2]
            req.roi.width = boundingbox[3]
            val res = ResponseFuture<ToggleCFtldTrackingWithBBResponse>()
            client.call(req, res)
            return res.toBooleanFuture()
        }
        throw RosRuntimeException("client error")

    }

    override fun startTracking(lastPose: Point3D?, threshold: Double?): Future<Boolean>? {
        TODO("Not yet implemented")
    }

    override fun stopTracking() {
        val req = clientTrigger!!.newMessage()
        //dont set data
        val res = ResponseFuture<ToggleCFtldTrackingWithBBResponse>()
        clientTrigger!!.call(req, res)
    }
}